mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala

```diff
@@ -169,6 +181,33 @@ class LinearRegression(override val uid: String)
     val yMean = ySummarizer.mean(0)
     val yStd = math.sqrt(ySummarizer.variance(0))
 
```

Comment:
1) `instances` will be cached, but if we go the `normal` path, `instances` will be never unpersisted. 
2) Please move the if condition to the top, and if the solver is `auto`, we just do a take(1) to get number of features so we don't run unnecessary `val (featuresSummarizer, ySummarizer) `. This can solve 1) as well.